### ISSUE TITLE
Help: Does graphite-api has to be running on carbon-cache nodes?

### ISSUE THREAD

**Issue by kkdk5535** [brutasse/graphite-api#169]
Hi @brutasse 
Following is my current setup, I have graphite-web & grafana on 2 servers and carbon-caches on 10 separate servers. I did setup `graphite-api` on graphite-web node and configured it run but it appears its unable to fetch any events using following `graphite-api.yaml` setting,

``` yaml
search_index: /opt/graphite/index
finders:
  - graphite_api.finders.whisper.WhisperFinder
functions:
  - graphite_api.functions.SeriesFunctions
  - graphite_api.functions.PieFunctions
whisper:
  directories:
    - /opt/graphite/storage/whisper
time_zone: "America/Los_Angeles"

carbon:
  hosts:
    - 1.2.3.4:7002
    - 4.5.6.7:7002
    - 8.9.10.11:7002
    - 12.13.14.15:7002
    - 16.17.18.19:7002
    - 20.21.22.23:7002
    - 24.25.26.27:7002
    - 28.29.30.31:7002
    - 32.33.34.35:7002
    - 36.37.38.39:7002
  timeout: 1
  retry_delay: 15
  carbon_prefix: carbon
  replication_factor: 1

allowed_origins:
  - .*
cache:
  type: redis
  redis_host: localhost
```

In output it shows when I try to run graphite-api using `NGINX` + `UWSGI`,

```
2016-06-09_07:34:57.68839 index_path='/opt/graphite/index' event='reading index data'
2016-06-09_07:34:57.68891 index_path='/opt/graphite/index' duration=0.0003199577331542969 total_entries=0 event='search index reloaded'
```

My runit `sv` file for graphite-api

``` bash
#!/bin/sh
exec 2>&1
exec /usr/local/bin/uwsgi \
--processes 2 \
--threads 2 \
--uwsgi-socket :8888 \
--module graphite_api.app:app \
--uid graphite --gid graphite \
--no-orphans --master \
--procname graphite-api \
--die-on-term \
--pidfile "/var/run/uwsgi-graphite-api.pid"
```

**Questions:**
- Can't I deploy graphite-api on separate nodes other than carbon-cache? I haven't seen anything in documentation with respect to this question.
- If `graphite-api` cannot read files on remote cache server then what is the use of carbon-cache settings inside the `graphite-api.yaml` file?
- In my environment having 10 carbon-cache nodes its not practical to use graphite-api if it has to be running locally on cache nodes. Please share alternate solution to run `graphite-api` on grafana & nginx node cluster. I cannot use shared volume in my case as it will add another layer of network latency.

**Comment by kartoffelheinz** [comment]
We need the exact same functionality and thought that using remote carbon cache in graphite-api was possible (because you can specify this in the configuration file). But it seems it doesn't work.

Any updates on this? It would be nice to have atleast some kind of documentation, so we know what specifying carbon hosts in graphite-api.yml actually does.

**Comment by brutasse** [comment]
Yes, it should definitely be possible. Would it be possible to share your carbon-relay & graphite-api configuration?

It'd also be helpful to sniff network traffic on the carbon port to see if any request is being made.

**Comment by rlodge** [comment]
I'm having essentially the same problem, though my setup is simpler: two servers running carbon, and metrics distributed to them from carbon-relay-ng, and simple "carbon" configuration in graphite-api.yaml.

It seems to me like the carbonlink code only gets called after a local whisper directory has been found, so if the data exists entirely on a remote server, the graphite-api server doesn't know anything about the node I'm asking for, and just returns an empty array.

Is that how it's supposed to work?  (it's not clear to me if carbonlink will return stored data, or only currently-cached data anyway).  Or do we need a storage finder for a "remote graphite_api"?

**Comment by kkdk5535** [comment]
@brutasse 
Thanks for positive response, Following is my relay configuration on `carbon-cache` nodes,
```
[cache]
ENABLE_LOGROTATION = True
USER = graphite
MAX_CACHE_SIZE = inf
MAX_UPDATES_PER_SECOND = 85
MAX_UPDATES_PER_SECOND_ON_SHUTDOWN = 1000
MAX_CREATES_PER_MINUTE = 150
LINE_RECEIVER_INTERFACE = 127.0.0.1
PICKLE_RECEIVER_INTERFACE = 127.0.0.1
CACHE_QUERY_INTERFACE = 0.0.0.0
ENABLE_UDP_LISTENER = True
CACHE_WRITE_STRATEGY = naive
USE_FLOW_CONTROL = True
LOG_UPDATES = False
LOG_CACHE_HITS = False
WHISPER_AUTOFLUSH = False

[cache:a]
LINE_RECEIVER_PORT = 2003
UDP_RECEIVER_PORT = 2003
PICKLE_RECEIVER_PORT = 2004
CACHE_QUERY_PORT = 7002

[cache:b]
LINE_RECEIVER_PORT = 2005
UDP_RECEIVER_PORT = 2005
PICKLE_RECEIVER_PORT = 2006
CACHE_QUERY_PORT = 7003

[relay:a]
LINE_RECEIVER_INTERFACE = 0.0.0.0
LINE_RECEIVER_PORT = 2000
UDP_RECEIVER_PORT = 2000
MAX_QUEUE_SIZE = 10000000
PICKLE_RECEIVER_INTERFACE = 0.0.0.0
PICKLE_RECEIVER_PORT = 2001
RELAY_METHOD = consistent-hashing
DESTINATIONS = 127.0.0.1:2004:a, 127.0.0.1:2006:b
```

`graphite-api` config
```yaml
search_index: /opt/graphite/index
finders:
  - graphite_api.finders.whisper.WhisperFinder
functions:
  - graphite_api.functions.SeriesFunctions
  - graphite_api.functions.PieFunctions
whisper:
  directories:
    - /opt/graphite/storage/whisper
time_zone: "America/Los_Angeles"

carbon:
  hosts:
    - 1.2.3.4:7002
    - 4.5.6.7:7002
    - 8.9.10.11:7002
    - 12.13.14.15:7002
    - 16.17.18.19:7002
    - 20.21.22.23:7002
    - 24.25.26.27:7002
    - 28.29.30.31:7002
    - 32.33.34.35:7002
    - 36.37.38.39:7002
  timeout: 1
  retry_delay: 15
  carbon_prefix: carbon
  replication_factor: 1

allowed_origins:
  - .*
cache:
  type: redis
  redis_host: localhost
```

_It'd also be helpful to sniff network traffic on the carbon port to see if any request is being made._
Anything other than tcpdump would help? Any specific flags u want me to use when capturing tcpdump output?

**Comment by rlodge** [comment]
TCPDUMP showed that the carbonlink servers were all queried, but only once a metric was found; if a whisper file was on the local storage, the metric was returned and all carbonlink hosts were queried.  But the graphite api instance simply didn't know about whisper files on remote servers.

Since I'm not clear that carbonlink is the way this is supposed to work (I thought carbonlink was only to query not-yet-flushed cached metrics), but I wanted to distribute the load across multiple cache instances on multiple hosts, I created a [remote metric finder](https://gist.github.com/rlodge/d8f72f670da705b79eab348a531b50b1).  It seems to work fine, although I haven't tested it under load, and it requires that a graphite api instance be running on all nodes.  Each graphite api instance is configured with its own whisper directory, and with its own carbonlink setup for its local cache instance, and the one I use as my primary endpoint is configured as shown in the gist with a remote finder and remote host list.

**Comment by kkdk5535** [comment]
@rlodge 
Thanks for taking time and sharing details about your experience and [PATCH](https://gist.github.com/rlodge/d8f72f670da705b79eab348a531b50b1) u wrote to address the problem. I will test it and update you on my experience with your [PATCH](https://gist.github.com/rlodge/d8f72f670da705b79eab348a531b50b1)

**Comment by piotr1212** [comment]
I'm also trying to configure this but can't find where it is implemented. In graphite-web it is here: https://github.com/graphite-project/graphite-web/blob/master/webapp/graphite/remote_storage.py and configured through CLUSTER_SERVERS in local_settings.py

**Comment by olevchyk** [comment]
@kkdk5535 can you update what was the experience with the @rlodge 's  patch  or maybe someone found an alternative solution ?

**Comment by kkdk5535** [comment]
@olevchyk 
I was planning to try the PATCH contributed by @rlodge but I got busy with other projects. I am currently doing POC with https://prometheus.io

**Comment by olevchyk** [comment]
did few   tests and I can confirm the performance is really bad.  
Main problem is that we get raw data from remote /render HTTP endpoint and there is no way at this point you can limit number of points by specifying  maxDataPoints. 
only format=json supports maxDataPoints, tried to adjust this patch to get  json data instead of raw to be able to set  maxDataPoints but then you need to iterate throught the points to format response in specific for storage finders way (time_info, data-points)

If you have large number of metrics/ whisper database files then my advice - don't use this patch because with native setup you will got much better performance.

**Comment by kkdk5535** [comment]
@olevchyk 
Thanks for your time in testing and sharing your experience to community. I am going to close this issue as it appears no one is currently maintaining this project.
